File: app/src/main/java/com/numero/sojodia/adapter/BusTimePagerAdapter.java
package com.numero.sojodia.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;
import android.support.v4.view.PagerAdapter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.numero.sojodia.R;
import com.numero.sojodia.model.BusTime;

import java.util.List;

public class BusTimePagerAdapter extends PagerAdapter {

    private List<BusTime> busTimeList;
    private Resources resources;

    public BusTimePagerAdapter(Context context) {
        resources = context.getResources();
    }

    public BusTimePagerAdapter setBusTimeList(List<BusTime> busTimeList) {
        this.busTimeList = busTimeList;
        return this;
    }

    @SuppressLint("DefaultLocale")
    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        View view = LayoutInflater.from(container.getContext()).inflate(R.layout.pager_bus_time, null);

        TextView busTextView = (TextView) view.findViewById(R.id.bus_time_text);

        busTextView.setText(String.format("%s\n%02d:%02d", getBusType(position), busTimeList.get(position).hour, busTimeList.get(position).min));

        TextView nextBusTextView = (TextView) view.findViewById(R.id.next_bus_time_text);
        if (position == busTimeList.size() - 1) {
            nextBusTextView.setText(R.string.last_bus);
        } else {
            nextBusTextView.setText(String.format("Next\n%s\n%02d:%02d", getBusType(position + 1), busTimeList.get(position + 1).hour, busTimeList.get(position + 1).min));
        }

        container.addView(view);
        return view;
    }

    private String getBusType(int position) {
        if (busTimeList.get(position).isNonstop) {
            return resources.getString(R.string.bus_type_nonstop);
        }  else {
            return resources.getString(R.string.bus_type_regular);
        }
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
    }

    @Override
    public int getCount() {
        if (busTimeList == null) {
            return 0;
        }
        return busTimeList.size();
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }
}

File: app/src/main/kotlin/com/numero/sojodia/ui/board/BusTimePagerAdapter.kt
package com.numero.sojodia.ui.board

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import com.numero.sojodia.R
import com.numero.sojodia.databinding.PagerBusTimeBinding
import com.numero.sojodia.model.BusTime

class BusTimePagerAdapter : PagerAdapter() {

    private var busTimeList: List<BusTime>? = null

    fun setBusTimeList(busTimeList: List<BusTime>) {
        this.busTimeList = busTimeList
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val binding = PagerBusTimeBinding.inflate(LayoutInflater.from(container.context), container, false)

        val busTimeList = busTimeList ?: return binding.root

        if (position >= busTimeList.size) {
            return binding.root
        }

        val busTime = busTimeList[position]
        binding.busTimeTextView.text = "%02d:%02d".format(busTime.time.hour, busTime.time.min)
        binding.busDescriptionTextView.apply {
            if (busTime.isNonstop) {
                visibility = View.VISIBLE
                setText(R.string.bus_type_nonstop)
            } else {
                visibility = View.GONE
            }
            if (position == busTimeList.size - 1) {
                // 最終バス
                visibility = View.VISIBLE
                setText(R.string.last_bus)
            }
        }

        if (position != busTimeList.size - 1) {
            val nextBusTime = busTimeList[position + 1]
            binding.nextBusTimeTextView.text = "%02d:%02d".format(nextBusTime.time.hour, nextBusTime.time.min)
            binding.nextBusDescriptionTextView.apply {
                if (nextBusTime.isNonstop) {
                    visibility = View.VISIBLE
                    setText(R.string.bus_type_nonstop)
                } else {
                    visibility = View.GONE
                }
            }
        }

        container.addView(binding.root)
        return binding.root
    }

    override fun destroyItem(container: ViewGroup, position: Int, any: Any) {
        container.removeView(any as View)
    }

    override fun getCount(): Int {
        return busTimeList?.size ?: return 0
    }

    override fun isViewFromObject(view: View, any: Any): Boolean {
        return view === any
    }

    override fun getItemPosition(any: Any): Int {
        return PagerAdapter.POSITION_NONE
    }
}
